src/storage/src/render/mod.rs

```diff
@@ -144,105 +141,112 @@ pub fn build_storage_dataflow<A: Allocate, B: StorageCapture>(
+                        (),
+                        SourceData,
+                        mz_repr::Timestamp,
+                        mz_repr::Diff,
+                    >(
+                        source.storage_metadata.persist_location,
+                        source.storage_metadata.persist_shard,
+                    )
+                    .await
+                    .expect("could not open persist shard");
+
+                    while scheduler.notified().await {
+                        let frontier = frontiers.borrow()[0].clone();
+
+                        if !active_write_worker
+                            || weak_token.upgrade().is_none()
+                            || shared_frontier.borrow().is_empty()
+                        {
+                            return;
                         }
-                    })
-                    .as_collection();
-                    let err = source(region, "InvalidSource", |cap, info| {
-                        let mut act_cap = Some(ActivateCapability::new(
-                            cap,
-                            &info.address,
-                            region.activations(),
-                        ));
-
-                        let drop_activator = Arc::new(scope.sync_activator_for(&info.address[..]));
-                        let drop_activator_weak = Arc::downgrade(&drop_activator);
-
-                        tokens.push(drop_activator);
-
-                        move |_handle| {
-                            if drop_activator_weak.upgrade().is_some() {
-                                act_cap.take();
+
+                        while let Some((_cap, data)) = input.next() {
+                            data.swap(&mut buffer);
+
+                            for (row, ts, diff) in buffer.drain(..) {
+                                stash
+                                    .entry(ts)
+                                    .or_default()
+                                    .push((SourceData(row), ts, diff));
                             }
                         }
-                    })
-                    .as_collection();
-                    ((ok, err), Arc::new(tokens) as Arc<dyn Any + Send + Sync>)
-                };
 
-                boundary.capture(*src_id, ok, err, token, &debug_name, dataflow_id);
-            }
+                        let empty = Vec::new();
+                        let updates = stash
+                            .iter()
+                            .flat_map(|(ts, updates)| {
+                                if !frontier.less_equal(ts) {
+                                    updates.iter()
+                                } else {
+                                    empty.iter()
+                                }
+                            })
+                            .map(|&(ref row, ref ts, ref diff)| ((&(), row), ts, diff));
+
+                        if PartialOrder::less_than(&*shared_frontier.borrow(), &frontier) {
```

Comment:
My understanding is that this check fixes the over-agressive appending described here: https://materializeinc.slack.com/archives/C011P87EL2V/p1652887957780259?thread_ts=1652866088.494569&cid=C011P87EL2V right? might be worth a comment, especially as the difference between `shared_frontier` (an in-memory cache of the sources upper (?) read by`report_frontier_progress`) and `frontier` is somewhat unclear

(that reminds me, after this pr, we should write something about what `report_frontier_progress` actually does)

Reply:
`shared_frontier` is probably a bad name. Its value tracks the persist shard's upper, you can see it being updated right after this append call succeeds. `report_frontier_progress` simply sends asynchronous feedback to the storage controller about how far the upper has advanced

Reply:
`frontier` also had a bad name, because it was the input frontier of the data that are about to be persisted. I renamed the variable to `input_frontier`

Reply:
maybe name `shared_frontier` to `most_recent_persist_upper` or something like? for frontier i don't mind long, descriptive names